src/utils/deferNetworkRequests.ts

```diff
@@ -0,0 +1,38 @@
+/**
+ * Intercepts and defers any requests on the page
+ * until the Service Worker instance is ready.
+ */
+export function deferNetworkRequests(
+  workerReady: Promise<ServiceWorkerRegistration | null>,
+) {
+  // Defer `XMLHttpRequest` until the Service Worker is ready.
+  const originalXhrSend = window.XMLHttpRequest.prototype.send
+  const restoreXhrSend = function (
+    this: XMLHttpRequest,
+    ...args: Parameters<XMLHttpRequest['send']>
+  ) {
+    window.XMLHttpRequest.prototype.send = originalXhrSend
+    this.send(...args)
```

Comment:
I'm not sure if this is a problem or not, but here we're first restoring the `send` function, and then we're calling `send` on `this` (which would be the function we just restored). I find this confusing.

Is it possible that using `call` would be cleaner for this case?

```diff
- this.send(...args)
+ window.XMLHttpRequest.prototype.send.call(window.XMLHttpRequest, ...args)
```

Reply:
`this` in the current context is a reference to a created `XMLHttpRequest` instance, while binding it to `window.XMLHttpRequest` would bind it to the _class_.

I'm stubbing the _class method_, so it affects all instances that inherit that class and call that method. But when restoring, I restore the class method (what I stubbed), and call the `this.send()`, which means performing this current deferred request.

Reply:
Makes sense! Thanks for clarifying 🎉